pyomo/contrib/gdpbb/GDPbb.py

```diff
@@ -0,0 +1,252 @@
+            while len(heap) > 0:
+                # pop best model off of heap
+                sort_tup, mdl, mdl_results, vars = heapq.heappop(heap)
+                old_obj_val, disjunctions_left, _ = sort_tup
+                config.logger.info("Exploring node with LB %.10g and %s inactive disjunctions." % (
+                    old_obj_val, disjunctions_left
+                ))
+
+                # if all the originally active disjunctions are active, solve and
+                # return solution
+                if disjunctions_left == 0:
+                    config.logger.info("Model solved.")
+                    # Model is solved. Copy over solution values.
+                    for orig_var, soln_var in zip(model.GDPbb_utils.variable_list, vars):
+                        orig_var.value = soln_var.value
+
+                    solve_data.results.problem.lower_bound = mdl_results.problem.lower_bound
+                    solve_data.results.problem.upper_bound = mdl_results.problem.upper_bound
+                    solve_data.results.solver.timing = solve_data.timing
+                    solve_data.results.solver.termination_condition = mdl_results.solver.termination_condition
+                    return solve_data.results
+
+                next_disjunction = mdl.GDPbb_utils.unenforced_disjunctions.pop(0)
+                config.logger.info("Activating disjunction %s" % next_disjunction.name)
+                next_disjunction.activate()
+                mdl.GDPbb_utils.curr_active_disjunctions.append(next_disjunction)
+                djn_left = len(mdl.GDPbb_utils.unenforced_disjunctions)
+                for disj in next_disjunction.disjuncts:
+                    disj._activate_without_unfixing_indicator()
+                    if not disj.indicator_var.fixed:
+                        disj.indicator_var = 0  # initially set all indicator vars to zero
+                for disj in next_disjunction.disjuncts:
+                    if not disj.indicator_var.fixed:
+                        disj.indicator_var = 1
+                    mnew = mdl.clone()
+                    if not disj.indicator_var.fixed:
+                        disj.indicator_var = 0
+                    obj_value, result, vars = self.subproblem_solve(mnew, solver, config)
+                    counter += 1
+                    ordering_tuple = (obj_sign * obj_value, djn_left, -counter)
+                    heapq.heappush(heap, (ordering_tuple, mnew, result, vars))
+                config.logger.info("Added %s new nodes with %s relaxed disjunctions to the heap. Size now %s." % (
+                    len(next_disjunction.disjuncts), djn_left, len(heap)))
+
+    def validate_model(self, model):
+        # Validates that model has only exclusive disjunctions
+        for d in model.component_data_objects(
+                ctype=Disjunction, active=True):
+            if (not d.xor):
+                raise ValueError('GDPlbb unable to handle '
```

Comment:
What is the 'L' in GDPlbb'?

Reply:
Holdover from previous name of solver. Has been resolved